inference/primitive.go

```diff
@@ -38,67 +42,300 @@ type primitiveFullTrigger struct {
+// nilabilities of upstream sites in the imported InferredMap since the positions of the primitive
+// sites are different. The primitivizer here contains logic to fix such discrepancies.
+// Note that this fix (mapping) is two-ways:
+// (1) Local incorrect upstream position to correct one for matching upstream sites in the
+// inferred map (such that all position information in primitiveSite is always correct). This is
+// done via primitivizer.site.
+// (2) Correct upstream position to local incorrect one for error reporting purposes (the analysis
+// framework only allows reporting via local token.Pos). This is done via primitivizer.sitePos.
+//
+// [archive importer]: https://github.com/golang/tools/blob/fa12f34b4218307705bf0365ab7df7c119b3653a/internal/gcimporter/bimport.go#L59-L69
+type primitivizer struct {
+	pass *analysis.Pass
+	// upstreamObjPositions maps "<pkg repr>.<object path>" to the correct position.
+	upstreamObjPositions map[string]token.Position
+	// files maps the file name (modulo the possible build-system prefix) to the token.File object
+	// for faster lookup when converting correct upstream position back to local token.Pos for
+	// reporting purposes.
+	files map[string]fileInfo
+	// curDir is the current working directory, which is used to trim the prefix (e.g., bazel
+	// random sandbox prefix) from the file names for cross-package references.
+	curDir string
+}
+
+// newPrimitivizer returns a new and properly-initialized primitivizer.
+func newPrimitivizer(pass *analysis.Pass) *primitivizer {
+	// To tackle the position discrepancies for upstream sites, we have added an ObjectPath field
+	// to primitiveSite, which can be used to uniquely identify an exported object relative to the
+	// package. Then, we simply cache the correct position information when importing
+	// InferredMaps, since the positions collected in the upstream analysis are always correct.
+	// Later when querying upstream objects in downstream analysis, we can look up the cache and
+	// fill in the correct position in the returned primitive site instead.
+
+	// Create a cache for upstream object positions.
+	upstreamObjPositions := make(map[string]token.Position)
+	for _, pkgFact := range pass.AllPackageFacts() {
+		importedMap, ok := pkgFact.Fact.(*InferredMap)
+		if !ok {
+			continue
+		}
+
+		importedMap.Range(func(site primitiveSite, _ InferredVal) bool {
+			if site.ObjectPath == "" {
+				return true
+			}
+
+			objRepr := site.PkgRepr + "." + string(site.ObjectPath)
+			if existing, ok := upstreamObjPositions[objRepr]; ok && existing != site.Position {
+				panic(fmt.Sprintf(
+					"conflicting position information on upstream object %q: existing: %v, got: %v",
+					objRepr, existing, site.Position))
+			}
+			upstreamObjPositions[objRepr] = site.Position
+			return true
+		})
+	}
+
+	// Find the current working directory (e.g., random sandbox prefix) for trimming the file names.
```

Comment:
Interesting, does bazel set the CWD to something inside `bazel-out/` regardless of where the `bazel build` command is run from? How about other build systems? Feels like, in general, making all paths relative to CWD for most build systems would cause issues when you run the build command from different directories...

Maybe we want to have this logic iff cwd includes `bazel-out/`? Or does that directory name change?

Reply:
Makes sense. I suspect one day we will run into yet another analysis driver that breaks this in some other way (buck.v2?), but it's a bridge we can cross when we get to it.